cilium/CFP-28427-fqdn-selectors.md

```diff
@@ -0,0 +1,409 @@
+
+However, all hope is not lost. The state of the FQDN policy engine is also checkpointed to disk, so we can mostly recreate the set of labels that apply to a given IP. Additionally, we will **checkpoint the local allocator state to disk**, so that the same set of labels receives the same identity.
+
+Checkpointing the local allocator is required so that policy maps (which contain only numeric IDs) remain stable across restarts.
+
+The full procedure is as follows:
+
+1. Read the most recent allocator state. This maps Labels -> Numeric Identity (NID). Insert this in the allocator.
+2. Read the IPCache state. This maps Prefix -> Numeric Identity
+3. For any NIDs in the IPCache and not in the allocator, use the "withold + request" mechanism to request the same NID again
+4. Wait the synchronization grace period (10 minutes)
+5. Release all restored references (identity allocations, placeholder ipcache metadata)
+
+In this proposal, the Label -> NID mapping is now fixed. Thus, for prefixes to have stable NIDs across restarts, they must have the same labels. The agent waits for its data sources to synchronize before computing labels, so this should primarily be the case.
+
+#### Example
+
+The IPCache has three prefixes, each with a corresponding identity.
+
+- `192.168.0.0/24` -> `(cidr:192.168.0.0/24)`
+- `1.1.1.1/32` -> `(fqdn:*, fqdn:one.one.one.one)`
+- `2.2.2.2/32` -> `(fqdn:*)`
+
+If, on agent restart, all prefixes have the same set of labels, then the identity mapping is stable. If, however, the FQDN checkpointing is not exact (which does happen when an IP is near expiration), then `1.1.1.1/32` may change labels to `(fqdn:*)`. However, since the identity for `fqdn:*` is already allowed by policy maps, no traffic to `fqdn:*` is dropped.
+
+Thus, transitioning to label-based, rather than prefix-based allocation should preserve policy stability on restarts.
+
+### Key question: Circuit breakers?
+
+There is currently a circuit-breaker, `--tofqdns-endpoint-max-ip-per-hostname`, which is intended to prevent names such as S3 causing identity churn. In this proposal, large names such as S3 should not allocate more than one or two identities, making this less critical.
+
+However, this is still a security-sensitive area, since we are allocating resources based on external, untrusted packets. We will need a sane circuit-breaker to ensure we do not allocate a large number of identities based on malicious input.
+
+The maximum number of identities allocated by ToFQDN selectors is the number of unique selector combinations seen for given IPs. The theoretical maximum is, therefore, $2^N-1$[^1]. 10 distinct selectors would be OK (1024). More than 16 selectors could *theoretically* exceed identity space.
```

Comment:
Do you think there's something here we could export as a metric to at least measure the risk / likelihood of running into an issue with this?

Reply:
We do have a metric for the total number of local identities. We can easily add an alert for agents with a large number of local identities. This would be a very good start.

Reply:
I see, that is probably good enough. I was wondering about this "multiple selectors selecting the same IP" aspect since that can end up generating double the number of identities for each new selector selecting the IP. But this is _probably_ getting too much into the weeds for metrics?

Reply:
From an end-user perspective, the interesting thing is memory usage and policy map size. It doesn't matter what leads to those identities, just that they're there -- 10,000 fqdn selectors vs. a pathological combination of 16 of them, that's all the same to the admin.

AIUI we already have alerts for overfull policy maps (I hope!). An alert for excessive local identities would be a good addition to this, I'll look in to adding this.

Reply:
I will add an optional follow-up item to think about if there are metrics we can add. Having some metric that indicates that the user chose selectors that are risking combinatorial explosion would be nice to have, as it would be a more targeted signal than "policy map is full" or "allocator is out of identities"